_posts/2018/2018-01-07-extract-conditionals-into-well-named-methods.md

```diff
@@ -0,0 +1,101 @@
+---
+title: 'Extract conditionals into well-named methods'
+description: 'An extraction to a private method often adds clarity'
+layout: article
+category: ruby
+image:
+  base: '2018/extract-conditionals-into-well-named-methods'
+  alt: 'Jigsaw pieces'
+  source: 'https://unsplash.com/photos/3y1zF4hIPCg'
+  credit: 'Hans-Peter Gauster'
+---
+
+One of the simplest (and most effective) ways to refactor is to [extract a method](https://refactoring.com/catalog/extractMethod.html). The naming of an extracted method is a great tool for communicating the thinking behind the code.
+
+## Instead of…
+
+...using slightly complex statements in your conditionals.
+
+```ruby
+class BrightonCoffeeShop
+  def initialize(name)
+    @name = name
+  end
+
+  def good?
+    if @name == 'Starbucks' || @name == 'Costa' ||
+       @name == 'Barry’s Caff'
+      false
+    elsif @name == 'Coffee@33' || @name == 'Small Batch'
+      true
+    else
+      false
+    end
+  end
+end
+```
+
+This implementation explicitly returns `false` if the `name` is unknown. The only way to guarantee a decent coffee shop experience is to rule out places you’ve never heard of.
+
+
+## Use…
+
+...the Extract Method refactoring to move the logic into a `private` method, giving that method a name that clearly explains the concept of the code.
+
+```ruby
+class BrightonCoffeeShop
+  def initialize(name)
+    @name = name
+  end
+
+  def good?
+    clean? && local? && excellent_coffee?
+  end
+
+  private
+
+  def clean?
+    !filthy?
+  end
+
+  def excellent_coffee?
+    ['Coffee@33', 'Small Batch'].include?(@name)
+  end
+
+  def filthy?
+    @name == 'Barry’s Caff'
+  end
+
+  def local?
+    !national_chain?
+  end
+
+  def national_chain?
+    ['Starbucks', 'Costa'].include?(@name)
+  end
+end
+```
+
+As part of our refactoring, we’re now using `Array#include?` for the lists of names.
+
+
+## But why?
+
+By extracting a method you, by definition, name it. Our method clearly states the specific criteria for a good coffee shop. These private methods act as clear documentation for the class.
+
+Before, all we really had was a list of 'good' and 'bad' coffee shops. We had no knowledge of how those lists were developed.
+
+A naive refactoring of this class might have just stuffed all the good and bad names in arrays named `GOOD` and `BAD` and checked the coffee shop `name` against them. We would have lost the reasoning behind _why_ the places were good or bad. And the _why_ is the most important part.
+
+In this case I've also extracted the negative concepts of `filthy?` and `national_chain?` alongside the positive concepts like `#clean?` and `#local?`. This avoids using the negative versions of the private methods in the `#good?` method. This may be overkill in some cases, but here it adds useful context as this class is heavily geared towards what makes a decent coffee shop.
```

Comment:
*heavily geared towards _the positive side of_ what makes a decent coffee shop...

Perhaps..? Because what makes a decent coffee shop is also that it's not filthy, it's not a local chain, etc..